openpype/hosts/maya/plugins/publish/validate_xgen.py

```diff
@@ -0,0 +1,64 @@
+import json
+
+import maya.cmds as cmds
+import xgenm
+
+import pyblish.api
+
+
+class ValidateXgen(pyblish.api.InstancePlugin):
+    """Validate Xgen data."""
+
+    label = "Validate Xgen"
+    order = pyblish.api.ValidatorOrder
+    host = ["maya"]
+    families = ["xgen"]
+
+    def process(self, instance):
+        # Validate only xgen collections are in objectset.
+        nodes = set(
+            instance.data["xgenNodes"] +
+            cmds.ls(instance, type="transform", long=True)
+        )
+        remainder_nodes = []
+        for node in instance:
+            if node in nodes:
+                continue
+            remainder_nodes.append(node)
+
+        msg = (
+            "Only the collection is used when publishing. Found these invalid"
+            " nodes in the objectset:\n{}".format(remainder_nodes)
+        )
+        assert not remainder_nodes, msg
```

Comment:
No skin off my nose, but why not use assertion?

Reply:
Not doing assertion was mostly because:
- the if statement helped to only format the string if the error occurred - it might actually do that too if the string formatting is moved to after the comma in the assertion, but I'm not entirely sure.
- to prepare the code for the new publisher which will require raising `KnownPublishError` errors as far as I know.
- because technically assertion errors can be disabled - [e.g. with `PYTHONOPTIMIZE`](https://realpython.com/python-assert-statement/#disabling-assertions-in-production-for-performance) environment variable. Also see [_When not to use assertions_](https://realpython.com/python-assert-statement/#when-not-to-use-assertions)